blockly-core/core/ui/contract_editor/contract_editor.js

```diff
@@ -394,15 +396,33 @@ Blockly.ContractEditor.prototype.moveExampleBlocksToModal_ = function (functionN
   }, this);
 };
 
+/**
+ * Allows app to pass in a function that should be run against an example block
+ * when user hits Test button
+ * @param {function}
+ */
 Blockly.ContractEditor.prototype.registerTestHandler = function(testHandler) {
   this.testHandler_ = testHandler;
 };
 
+/**
+ * Allows app to pass in a function that should be run when user resets a test
+ * @param {function}
+ */
 Blockly.ContractEditor.prototype.registerTestResetHandler = function (testResetHandler) {
   this.testResetHandler_ = testResetHandler;
 };
 
 /**
+ * Allows app to pass in a function that is called on contract editor close if
+ * any examples fail. This function should return true if the app wants to own
+ * closing the dialog
+ */
+Blockly.ContractEditor.prototype.registerTestsFailedOnCloseHandler = function (handler) {
+  this.appHijackedDialogClose_ = handler;
```

Comment:
Wondering what might be a good name for this to make its behavior a little more clear.

The callback kind of fills two roles, (1) return true/false if it should subsume the closing responsibility, (2) execute custom logic on a failure's close.

One option would be to, s/`this.appsHijackedDialogClose_/`this.customFailureCloseHandler_`, and if `this.customFailureCloseHandler_` is null, close normally on fail. Then the existence handles responsibility #1, and the callback has one less job.